File: trunk/editor/ARCed/tests/ARC_Data_serialization/arc_data_dump.rb
load 'data.rb'

module ARC

	module Data
	
		VERSION = "\x01\x00"
		
		DEBUG_DISABLE_STRING_MAPPING = false
		DEBUG_DISABLE_OBJECT_MAPPING = false
		
		TYPES = {
			NilClass => 0x10.chr,
			FalseClass => 0x11.chr,
			TrueClass => 0x12.chr,
			Fixnum => 0x21.chr,
			Bignum => 0x22.chr,
			Float => 0x23.chr,
			String => 0x30.chr,
			Array => 0x40.chr,
			Hash => 0x41.chr,
			Object => 0x00.chr
		}
		
		@@io = nil
		@@objects = [nil]
		@@class_path_redirects = {}
		
		def self.dump(io, obj, redirects = {})
			@@class_path_redirects = redirects
			@@io = io
			@@io.write(VERSION)
			begin
				self._dump(obj)
			rescue
				puts "stream position: #{@@io.tell}" rescue nil
				raise
			end
			self._reset
		end
		
		def self.load(io, redirects = {})
			@@class_path_redirects = redirects
			@@io = io
			version = @@io.read(2)
			raise "Error: #{self} version mismatch! Expected: #{VERSION.inspect} Found: #{version.inspect}" if VERSION != version
			begin
				data = self._load
			rescue
				puts "stream position: #{@@io.tell}" rescue nil
				raise
			end
			self._reset
			return data
		end
		
		def self._reset
			@@io = nil
			@@objects = [nil]
			@@class_path_redirects = {}
		end
  
		def self.__get_class_path(name)
			return (@@class_path_redirects.has_key?(name) ? @@class_path_redirects[name] : name)
		end
		
		def self.__try_map_object(obj)
			index = @@objects.index(obj)
			if index == nil
				self.__dump_int32(@@objects.size)
				@@objects.push(obj)
				return true
			end
			self.__dump_int32(index)
			return false
		end

		def self.__map_object(obj)
			@@objects.push(obj)
		end

		def self.__find_mapped_object(id)
			return (id < @@objects.size ? @@objects[id] : nil)
		end
		
		def self.__dump_int32(obj)
			@@io.write([obj].pack("V"))
		end

		def self.__load_int32
			return @@io.read(4).unpack("V")[0]
		end

		def self._dump(obj)
			return self._dump_nil(obj) if obj.is_a?(NilClass)
			return self._dump_false(obj) if obj.is_a?(FalseClass)
			return self._dump_true(obj) if obj.is_a?(TrueClass)
			return self._dump_fixnum(obj) if obj.is_a?(Fixnum)
			return self._dump_bignum(obj) if obj.is_a?(Bignum)
			return self._dump_float(obj) if obj.is_a?(Float)
			return self._dump_string(obj) if obj.is_a?(String)
			return self._dump_array(obj) if obj.is_a?(Array)
			return self._dump_hash(obj) if obj.is_a?(Hash)
			return self._dump_object(obj) if obj.is_a?(Object)
			raise "Error: #{obj.class} cannot be dumped!"
		end
		
		def self._load
			type = @@io.read(1)
			case type
			when TYPES[NilClass] then return self._load_nil
			when TYPES[FalseClass] then return self._load_false
			when TYPES[TrueClass] then return self._load_true
			when TYPES[Fixnum] then return self._load_fixnum
			when TYPES[Bignum] then return self._load_bignum
			when TYPES[Float] then return self._load_float
			when TYPES[String] then return self._load_string
			when TYPES[Array] then return self._load_array
			when TYPES[Hash] then return self._load_hash
			when TYPES[Object] then return self._load_object
			end
			raise "Error: Unknown type 0x%02X detected!" % type.ord
		end
		
		def self._dump_nil(obj)
			@@io.write(TYPES[NilClass])
		end

		def self._dump_false(obj)
			@@io.write(TYPES[FalseClass])
		end

		def self._dump_true(obj)
			@@io.write(TYPES[TrueClass])
		end
		
		def self._dump_fixnum(obj)
			@@io.write(TYPES[Fixnum])
			self.__dump_int32(obj)
		end
		
		def self._dump_bignum(obj)
			@@io.write(TYPES[Bignum])
			self.__dump_int32(obj) # our C++ implementation uses a "long" of 32 bit
		end
		
		def self._dump_float(obj)
			@@io.write(TYPES[Float])
			@@io.write([obj].pack("e"))
		end
		
		def self._dump_string(obj)
			@@io.write(TYPES[String])
			if obj.size > 0
				if !DEBUG_DISABLE_STRING_MAPPING && !DEBUG_DISABLE_OBJECT_MAPPING
					return if !self.__try_map_object(obj) # abort if object has already been mapped
				end
				self.__dump_int32(obj.size)
				@@io.write(obj)
			else
				self.__dump_int32(0)
			end
		end
		
		def self._dump_array(obj)
			@@io.write(TYPES[Array])
			if obj.size > 0
				if !DEBUG_DISABLE_OBJECT_MAPPING
					return if !self.__try_map_object(obj) # abort if object has already been mapped
				end
				self.__dump_int32(obj.size)
				obj.each {|value| self._dump(value)}
			else
				self.__dump_int32(0)
			end
		end
		
		def self._dump_hash(obj)
			@@io.write(TYPES[Hash])
			if obj.size > 0
				if !DEBUG_DISABLE_OBJECT_MAPPING
					return if !self.__try_map_object(obj) # abort if object has already been mapped
				end
				self.__dump_int32(obj.size)
				obj.each_pair {|key, value| self._dump(key); self._dump(value)}
			else
				self.__dump_int32(0)
			end
		end
		
		def self._dump_object(obj)
			@@io.write(TYPES[Object])
			self._dump_string(self.__get_class_path(obj.class.name)) # first the string path because this is required to load the object
			if !DEBUG_DISABLE_OBJECT_MAPPING
				return if !self.__try_map_object(obj) # abort if object has already been mapped
			end
			if obj.respond_to?("_arc_dump")
				data = obj._arc_dump
				self.__dump_int32(data.size)
				@@io.write(data)
			else
				variables = obj.instance_variables
				self.__dump_int32(variables.size)
				variables.each {|variable|
					self._dump_string(variable.to_s.gsub("@", ""))
					self._dump(obj.instance_variable_get(variable))
				}
			end
		end
		
		def self._load_nil
			return nil
		end

		def self._load_false
			return false
		end

		def self._load_true
			return true
		end
		
		def self._load_fixnum
			return self.__load_int32
		end
		
		def self._load_bignum
			return self.__load_int32 # our C++ implementation uses a "long" of 32 bit
		end
		
		def self._load_float
			return @@io.read(4).unpack("e")[0]
		end
		
		def self._load_string
			if !DEBUG_DISABLE_STRING_MAPPING && !DEBUG_DISABLE_OBJECT_MAPPING
				id = self.__load_int32
				return "" if id == 0
				obj = self.__find_mapped_object(id)
				return obj.clone if obj != nil
			end
			size = self.__load_int32
			obj = @@io.read(size)
			if !DEBUG_DISABLE_STRING_MAPPING && !DEBUG_DISABLE_OBJECT_MAPPING
				self.__map_object(obj)
			end
			return obj
		end
		
		def self._load_array
			if !DEBUG_DISABLE_OBJECT_MAPPING
				id = self.__load_int32
				return [] if id == 0
				obj = self.__find_mapped_object(id)
				return obj if obj != nil
			end
			size = self.__load_int32
			obj = []
			if !DEBUG_DISABLE_OBJECT_MAPPING
				self.__map_object(obj)
			end
			size.times {obj.push(self._load)}
			return obj
		end
		
		def self._load_hash
			if !DEBUG_DISABLE_OBJECT_MAPPING
				id = self.__load_int32
				return {} if id == 0
				obj = self.__find_mapped_object(id)
				return obj if obj != nil
			end
			size = self.__load_int32
			obj = {}
			if !DEBUG_DISABLE_OBJECT_MAPPING
				self.__map_object(obj)
			end
			size.times {key = self._load; obj[key] = self._load} # making sure key is always loaded first
			return obj
		end
		
		def self._load_object
			class_path = self._load
			if !DEBUG_DISABLE_OBJECT_MAPPING
				obj = self.__find_mapped_object(self.__load_int32)
				return obj if obj != nil
			end
			classes = class_path.split("::")
			if !Kernel.const_defined?(classes[0].to_sym)
				raise TypeError, "Class not defined: #{classes[0]}"
			end
			classe = Kernel.const_get(classes.shift.to_sym)
			classes.each {|c|
				if !classe.const_defined?(c.to_sym)
					raise TypeError, "Class not defined: #{c}"
				end
				classe = classe.const_get(c.to_sym)
			}
			size = self.__load_int32
			if classe.respond_to?("_arc_load")
				obj = classe._arc_load(@@io.read(size))
				if !DEBUG_DISABLE_OBJECT_MAPPING
					self.__map_object(obj)
				end
				return obj
			end
			obj = classe.allocate
			if !DEBUG_DISABLE_OBJECT_MAPPING
				self.__map_object(obj)
			end
			size.times {obj.instance_variable_set(("@" + self._load).to_sym, self._load)}
			return obj
		end
		
	end
	
end

File: trunk/editor/ARCed/tests/ARC_Data_serialization/arc_data_dump.py
from struct import pack, unpack
import types

class Data(object):

    _VERSION = "\x01\x00" # 1.0
    
    _TYPES = {
        "NoneClass" : chr(0x10),
        "FalseClass" : chr(0x11),
        "TrueClass" : chr(0x12),
        "Fixnum" : chr(0x21),
        "Bignum" : chr(0x22),
        "Float" : chr(0x23),
        "String" : chr(0x30),
        "Array" : chr(0x40),
        "Hash" : chr(0x41),
        "Object" : chr(0x00)
        }
    
    _io = None
    _strings = [None]
    _arrays = [None]
    _hashes = [None]
    _objects = [None]
    _class_path_redirects = {}
    _extended_namespace = {}
    
    @staticmethod
    def dump(io, obj, redirects = {}):
        global _class_path_redirects
        global _io
        _class_path_redirects = redirects
        _io = io
        _io.write(VERSION)
        try:
            _dump(obj)
        except Exception:
            try:
                print "stream position: %s" % _io.tell()
            except Exception:
                pass
            raise
        _reset()
    
    
    @staticmethod
    def load(io, redirects = {}, extended_namespace={}):
        global _class_path_redirects
        global _io
        global _extended_namespace
        _class_path_redirects = redirects
        _extended_namespace = dict(extended_namespace.update(globals()))
        extended_namespace = {}
        _io = io
        version = _io.read(2)
        if _VERSION != version:
            raise "Error: version mismatch! Expected: %s Found: %s" %(repr(VERSION), repr(version))
        try:
            data = _load()
        except Execption:
            try:
                print "stream position: %s" % _io.tell()
            except Exception:
                pass
            raise
        _reset()
        return data
    
    
    @staticmethod
    def _reset():
        global _io
        global _strings
        global _arrays
        global _hashes
        global _objects
        global _class_path_redirects
        global _extended_namespace
        _io = None
        _strings = [None]
        _arrays = [None]
        _hashes = [None]
        _objects = [None]
        _class_path_redirects = {}
        _extended_namespace = {}
    

    @staticmethod
    def __get_class_path(name):
        if _class_path_redirects.has_key(name):
            return _class_path_redirects[name]
        else: 
            return name
    
    
    @staticmethod
    def __get_class_object(class_path):
        classes = class_path.split("::")
        if not _extended_namespace.has_key(classes[0]):
            raise TypeError("Class not defined: %s" % classes[0])    
        classe = _extended_namespace[classes.pop(0)]
        for c in classes:
            if not classe.__dict__.has_key(c):
                raise TypeError("Class not defined: %s" % c)
            classe = classe.const_get(c.to_sym)
        return classe
    

    @staticmethod
    def __try_map(data, obj):
        index = data.index(obj)
        if index is None:
            __dump_int32(len(data))
            data.append(obj)
            return True
        __dump_int32(index)
        return False
    

    @staticmethod
    def __map(data, obj):
        data.append(obj)
    

    @staticmethod
    def __find_mapped(data, id):
        if id < len(data):
            return data[id]
        else:
            return None
    
    
    @staticmethod
    def __dump_int32(obj):
        _io.write(pack("<I", obj))
    

    @staticmethod
    def __load_int32():
        return _io.read(4).unpack("V")[0]
    
    
    @staticmethod
    def _dump(obj):
        if obj is None:
            return _dump_none(obj)
        if obj is False:
            return _dump_false(obj) 
        if obj is True:
            return _dump_true(obj)
        if isinstance(obj, types.IntType):
            return _dump_fixnum(obj)
        if isinstance(obj, types.LongType):
            return _dump_bignum(obj)
        if isinstance(obj, types.FloatType):
            return _dump_float(obj)
        if isinstance(ojb, types.StringTypes):
            return _dump_string(obj)
        if isinstance(obj, (types.ListType, types.TupleType)):
            return _dump_array(obj)
        if isinstance(obj, types.DictType):
            return _dump_hash(obj)
        if isinstance(obj, object):
            return _dump_object(obj)
        raise TypeError("Error: %s cannot be dumped!" % obj.__class__)
    
    @staticmethod
    def _load():
        type = _io.read(1)
        if type is _TYPES["NoneClass"]:
            return _load_none()
        elif type is _TYPES["FalseClass"]:
            return _load_false()
        elif type is _TYPES["TrueClass"]:
            return _load_true()
        elif type is _TYPES["Fixnum"]:
            return _load_fixnum()
        elif type is _TYPES["Bignum"]:
            return _load_bignum()
        elif type is _TYPES["Float"]:
            return _load_float()
        elif type is _TYPES["String"]:
            return _load_string()
        elif type is _TYPES["Array"]:
            return _load_array()
        elif type is _TYPES["Hash"]:
            return _load_hash()
        elif type is _TYPES["Object"]:
            return _load_object()
        
        raise "Error: Unknown type 0x%02X detected!" % type.ord
    
    
    @staticmethod
    def _dump_none(obj):
        _io.write(_TYPES["NoneClass"])
    

    @staticmethod
    def _dump_false(obj):
        _io.write(_TYPES["FalseClass"])
    

    @staticmethod
    def _dump_true(obj):
        _io.write(_TYPES["TrueClass"])
    
    
    @staticmethod
    def _dump_fixnum(obj):
        _io.write(_TYPES["Fixnum"])
        __dump_int32(obj)
    
    
    @staticmethod
    def _dump_bignum(obj):
        _io.write(_TYPES["Bignum"])
        __dump_int32(obj) # our C++ implementation uses a "long" of 32 bit
    
    
    @staticmethod
    def _dump_float(obj):
        _io.write(_TYPES["Float"])
        _io.write(pack("<f", obj))
    
    
    @staticmethod
    def _dump_string(obj):
        _io.write(_TYPES["String"])
        if obj.size > 0:
            if not __try_map(_strings, obj): # abort if object has already been mapped
                return
            __dump_int32(len(obj))
            _io.write(obj)
        else:
            __dump_int32(0)
        
    
    
    @staticmethod
    def _dump_array(obj):
        _io.write(_TYPES["Array"])
        if obj.size > 0:
            if not __try_map(_arrays, obj): # abort if object has already been mapped
                return 
            __dump_int32(len(obj))
            for value in obj:
                _dump(value)
        else:
            __dump_int32(0)
        
    
    
    @staticmethod
    def _dump_hash(obj):
        _io.write(_TYPES["Hash"])
        if obj.size > 0:
            if not __try_map(_hashes, obj): # abort if object has already been mapped
                return 
            __dump_int32(len(obj))
            for key, value in obj.items():
                _dump(key)
                _dump(value)
        else:
            __dump_int32(0)
        
    
    @staticmethod
    def _request_instance_variables(obj):
        if hasattr(obj, "_arc_instance_variables"):
            return obj._arc_instance_variables
        else:
            list = []
            for key, value in obj.__dict__.items():
                if key[0] is not "_":
                    if not isinstance(value, (types.FunctionType, types.ClassType, types.MethodType, 
                                              types.ModuleType, types.SliceType, types.LambdaType, 
                                              types.GeneratorType)):
                        list.append(key)
            return list
    
    @staticmethod
    def _dump_object(obj):
        _io.write(_TYPES["Object"])
        if hasattr(obj, "_arc_class_path"):
            klass_path = obj._arc_class_path
        else:
            klass_path = "%s::%s" % (obj.__class__.__module__, obj.__class__.__name__)
        _dump_string(__get_class_path(klass_path)) # first the string path because this is required to load the object
        if not __try_map(_objects, obj): # abort if object has already been mapped
            return
        if hasattr(obj, "_arc_dump"):
            data = obj._arc_dump()
            __dump_int32(len(data))
            _io.write(data)
        else:
            if hasattr(obj, "_arc_exclude"):
                excludes = obj._arc_exclude
            else:
                excludes = []
            variables = list(set(_request_instance_variables(obj)) - set(excludes))
            __dump_int32(len(variables))
            for variable in variables:
                _dump_string(variable)
                _dump(getattr(obj, variable))
        
    
    
    @staticmethod
    def _load_none():
        return None
    

    @staticmethod
    def _load_false():
        return False
    

    @staticmethod
    def _load_true():
        return True
    
    
    @staticmethod
    def _load_fixnum():
        return __load_int32()
    
    
    @staticmethod
    def _load_bignum():
        return __load_int32() # our C++ implementation uses a "long" of 32 bit
    
    
    @staticmethod
    def _load_float():
        return unpack("<f", _io.read(4))
    
    
    @staticmethod
    def _load_string():
        id = __load_int32()
        if id == 0:
            return ""
        obj = __find_mapped(_strings, id)
        if obj != None:
            return obj.clone
        size = __load_int32()
        obj = _io.read(size)
        __map(_strings, obj)
        return obj
    
    
    @staticmethod
    def _load_array():
        id = __load_int32()
        if id == 0:
            return []
        obj = __find_mapped(_arrays, id)
        if obj != None:
            return obj
        size = __load_int32()
        obj = []
        __map(_arrays, obj)
        for i in range(size):
            obj.append(_load())
        return obj
    
    
    @staticmethod
    def _load_hash():
        id = __load_int32()
        if id == 0:
            return {}
        obj = __find_mapped(_hashes, id)
        if obj != None:
            return obj
        size = __load_int32()
        obj = {}
        __map(_hashes, obj)
        for i in range(size):
            key = _load() # making sure key is always loaded first
            obj[key] = _load()
        return obj
    
    
    @staticmethod
    def _load_object():
        class_path = _load()
        obj = __find_mapped(_objects, __load_int32())
        if obj != None:
            return obj
        classe = __get_class_object(class_path)
        size = __load_int32()
        if hasattr(classe, "_arc_load"):
            obj = classe._arc_load(_io.read(size))
            __map(_objects, obj)
            return obj
        obj = classe.__new__(classe)
        __map(_objects, obj)
        for i in range(size):
            setattr(obj, _load(), _load())
        return obj
    
    



